cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTable.java

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.nosql;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.lib.AbstractCloseableIterator;
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.spi.data.InvalidFieldException;
+import co.cask.cdap.spi.data.StructuredRow;
+import co.cask.cdap.spi.data.StructuredTable;
+import co.cask.cdap.spi.data.table.StructuredTableSchema;
+import co.cask.cdap.spi.data.table.field.Field;
+import co.cask.cdap.spi.data.table.field.FieldFactory;
+import co.cask.cdap.spi.data.table.field.FieldType;
+import co.cask.cdap.spi.data.table.field.Range;
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Nosql structured table implementation.
+ */
+public final class NoSqlStructuredTable implements StructuredTable {
+  private static final Logger LOG = LoggerFactory.getLogger(NoSqlStructuredTable.class);
+  private final Table table;
+  private final StructuredTableSchema schema;
+  private final FieldFactory fieldFactory;
+
+  public NoSqlStructuredTable(Table table, StructuredTableSchema schema) {
+    this.table = table;
+    this.schema = schema;
+    this.fieldFactory = new FieldFactory(schema);
+  }
+
+  @Override
+  public void upsert(Collection<Field<?>> fields) throws InvalidFieldException {
+    LOG.trace("Table {}: Write fields {}", schema.getTableId(), fields);
+    Put put = convertFieldsToBytes(fields);
+    table.put(put.key, put.columns, put.values);
+  }
+
+  @Override
+  public Optional<StructuredRow> read(Collection<Field<?>> keys) throws InvalidFieldException {
+    return read(keys, Collections.emptySet());
```

Comment:
emptySet means all columns? I am not a fan of that semantics. It gets really confusing when a client computes the set of fields dynamically. Then they would have to check whether the set of fields ids empty and then skip the call. 

It is less critical for a get, but definitely a problem for updates or deletes. (if deleting an empty set of columns deletes all columns). 

Can we change this? Better use an optional or a similar way to have a separate value to indicate "all"

Reply:
I made the same comment in the original SPI PR... https://github.com/cdapio/cdap/pull/10874#discussion_r242242711

for the `read(keys, Collection)` method, we shouldn't allow empty column set

Reply:
Changed the read method to disallow empty columns passed in.